pkg/rsstorage/servers/file/server.go

```diff
@@ -127,28 +139,59 @@ func (s *StorageServer) CalculateUsage() (types.Usage, error) {
 
 // diskUsage will walk the specified path in a filesystem and
 // aggregate the size of the contained files.
-func diskUsage(duPath string, cacheTimeout time.Duration) (size datasize.ByteSize, err error) {
-	timeout := time.Now().Add(cacheTimeout)
-	sizep := &size
+func diskUsage(duPath string, cacheTimeout, walkTimeout time.Duration) (size datasize.ByteSize, err error) {
+	sizeChan := make(chan datasize.ByteSize)
+	errChan := make(chan error)
 
-	err = filepath.Walk(duPath, func(path string, info os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
+	go func(sizeChan chan datasize.ByteSize, errChan chan error) {
+		// prevent infinite loops
+		visited := map[string]interface{}{}
 
-		if time.Now().After(timeout) {
-			return errors.New("timeout in diskUsage")
-		}
+		err = filepath.Walk(duPath, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
 
-		if !info.IsDir() {
-			*sizep += datasize.ByteSize(info.Size())
-		}
+			if _, ok := visited[path]; ok {
+				return fmt.Errorf("diskUsage: encountered loop while walking directory, filepath: %s", path)
+			}
 
-		return nil
-	})
-	size = *sizep
+			visited[path] = true
 
-	return
+			if !info.IsDir() {
+				sizeChan <- datasize.ByteSize(info.Size())
+			}
+
+			return nil
+		})
+
+		close(sizeChan)
+		close(errChan)
+	}(sizeChan, errChan)
+
+	cacheTimeoutTimer := time.NewTimer(cacheTimeout)
+	defer cacheTimeoutTimer.Stop()
+
+	start := time.Now()
+
+	for {
+		select {
+		case <-cacheTimeoutTimer.C:
+			return 0, cacheTimeoutErr
+		case sz := <-sizeChan:
+			size += sz
+			start = time.Now()
+		// Success case error will return `nil`
+		case err = <-errChan:
+			return
+		default:
+			time.Sleep(walkCheckTime)
+		}
+
+		if time.Now().Sub(start) > walkTimeout {
+			return 0, walktimeoutErr
```

Comment:
If we return due to a `cacheTimoutTimer` tick or `walkTimeout`, should we force the filepath walk goroutine to exit?